### ISSUE TITLE
Changes for RC 40 as discussed at the MPI forum Dec 7-10, 2020 for Chapters 1, 2, 8, 16-18, A, B & others

### ISSUE THREAD

**Issue by RolfRabenseifner** [mpi-forum/mpi-issues#405]
# Problem

Summary of all changes reported by the reviewers for mainly Chapters 2, 8, 16-18, A, B and summarized in the attached pdf.

# Proposal

As voted on the RC 40, the changes shall be stored inseveral related pull requests

# Changes to the Text

See pdf and pull requests

Latest **commented pdf** with all changes from MPI-4.0 RC-1 (Dec. 2020) to RC-2 (Feb. 2021) 
that are related to this issue: 

> **[mpi40-rc-feb-21_rab.pdf](https://github.com/mpi-forum/mpi-issues/files/6050186/mpi40-rc-feb-21_rab.pdf) **
> and especially for PR547 MPI_Sessions_finalize: **[mpi40-issue435-PR547-sessions-2021-FEB-26-v1.pdf](https://github.com/mpi-forum/mpi-issues/files/6051904/mpi40-issue435-PR547-sessions-2021-FEB-26-v1.pdf)**

# List of latest pull requests that should be added to MPI-4.0 RC 2:

**Both, base PR and update PR not yet voted in:**

- [PR528](https://github.com/mpi-forum/mpi-standard/pull/258) to be updated with  [PR534](https://github.com/mpi-forum/mpi-standard/pull/534) - in Chap. 2 Terms
- [PR484](https://github.com/mpi-forum/mpi-standard/pull/484) to be updated with  [PR533](https://github.com/mpi-forum/mpi-standard/pull/533) - in Chap. 7 Groups, Contexts
- [PR434](https://github.com/mpi-forum/mpi-standard/pull/434) to be updated with  [PR544](https://github.com/mpi-forum/mpi-standard/pull/544) - in Chap. 18 Semantic changes (formerly Backward incompatibilities)  (+ not updated Chap. 19 and Annex B)
- [PR444](https://github.com/mpi-forum/mpi-standard/pull/444) to be updated with  [PR528](https://github.com/mpi-forum/mpi-standard/pull/528) - in Annex A.1
- [PR455](https://github.com/mpi-forum/mpi-standard/pull/) to be updated with  [PR425](https://github.com/mpi-forum/mpi-standard/pull/425) and  [PR540]( https://github.com/mpi-forum/mpi-standard/pull/540) - in Chap. 14 I/O (changes by Tony)

**For the following updates, the base is already voted in:**

- [PR351](https://github.com/mpi-forum/mpi-standard/pull/351) to be updated with  [PR531](https://github.com/mpi-forum/mpi-standard/pull/531) - in Chap. 5 Datatypes 
- [PR429](https://github.com/mpi-forum/mpi-standard/pull/429) to be updated with  [PR529](https://github.com/mpi-forum/mpi-standard/pull/529) - in Annex B Change-Log 
- ([PR412](https://github.com/mpi-forum/mpi-standard/pull/412)) to be updated with  [PR530](https://github.com/mpi-forum/mpi-standard/pull/530) - in Chap. 19 Language Bindings 
- [Issue 341](https://github.com/mpi-forum/mpi-issues/issues/341) to be updated with  [PR532](https://github.com/mpi-forum/mpi-standard/pull/532) - in Chap. 6 Collectives

**Additional changes also reported in the pdf**

- [Issue 437](https://github.com/mpi-forum/mpi-issues/issues/437) to be updated with  [PR517](https://github.com/mpi-forum/mpi-standard/pull/517) - in Chap. 4 Partitioned Communication
   (Additional question: is MPI_Parrived required to make progress? Answer from the Forum meeting Feb.2021: YES and change name to MPI_Ptest)

- [Issue 435](https://github.com/mpi-forum/mpi-issues/issues/435) to be updated with  [PR547](https://github.com/mpi-forum/mpi-standard/pull/547) - in Chap. 11.3 Sessions Model

# Impact on Implementations

None - all are corrections to the latex text only to remove errors or inconsistencies. 

# Impact on Users

None.

# Open, merged, and other issues/PR related to this issue 405

In the latest commented pdf (see below), the following open issues are reported (see "#-TODO" and "Done" comments in the pdf):
* [x] Merge of PR427 https://github.com/mpi-forum/mpi-standard/pull/427 Sorting of the changelog
* [x] Merge of PR434 https://github.com/mpi-forum/mpi-standard/pull/434  Corrections to Chap. 18, 19, and Annex B
* [x] Merge of PR439 https://github.com/mpi-forum/mpi-standard/pull/439 Corrections to Chap. 8 Topologies
* [x] Merge of PR443 https://github.com/mpi-forum/mpi-standard/pull/442 Corrections to Chap.10 The Info Object
* [x] Merge of PR442 https://github.com/mpi-forum/mpi-standard/pull/442 Corrections to Chap.16 Deprecated Interfaces
* [x] Merge of PR441 https://github.com/mpi-forum/mpi-standard/pull/441 Corrections to Chap.17 Removed Interfaces
* [x] Merge of PR444 https://github.com/mpi-forum/mpi-standard/pull/444 Corrections to Annex A. Language Binding Summaries
* [x] Merge of PR449 https://github.com/mpi-forum/mpi-standard/pull/449 Corrections to the Callback Prototype Function Index and some smaller ones in other chapters
* [ ] Additional modifications of the PR449 mpi-user-macs.tex to better fit to the existing concept.
* [x] Merge of PR461 https://github.com/mpi-forum/mpi-standard/pull/461 Corrections to the Change-Log
* [x] Merge of PR486 https://github.com/mpi-forum/mpi-standard/pull/486 Additional primary Fortran entries in the Declaration Index
* [x] Merge of PR497 https://github.com/mpi-forum/mpi-standard/pull/497 Additional changes to appLang-Const.tex according to Issue #375
* [x] Merge of PR498 https://github.com/mpi-forum/mpi-standard/pull/498 Additional changes to deprecated.tex according to Issue #366
* [x] Merge of PR499 https://github.com/mpi-forum/mpi-standard/pull/499 Adding " !(_c)" postfix embiggened mpi_f08 bindings according to Issue #364
* [x] Merge of PR493 https://github.com/mpi-forum/mpi-standard/pull/493 Secondary callback index entries at MPI...CREATE... routines
* [x] Merge of PR500 https://github.com/mpi-forum/mpi-standard/pull/500 Problem with 2-d arguments reported in https://github.com/mpi-forum/mpi-issues/issues/364#issuecomment-761755916
* [x] Merge of PR501 https://github.com/mpi-forum/mpi-standard/pull/501 Solving MPI_CONVERSION_FN_NULL embiggening problem in Issue #432
* [x] Merge of PR504 https://github.com/mpi-forum/mpi-standard/pull/504 Solving Annex A problems listed in Issue #332
* [x] Merge of PR506 https://github.com/mpi-forum/mpi-standard/pull/506 Solving Annex A final problems listed in Issue #332
* [x] Merge of PR508 https://github.com/mpi-forum/mpi-standard/pull/508 Small corrections to Chap. 1, listed in Issue #349

Related Pull Requests from other authors (see also "#-TODO" and "Done" comments in the latest commented pdf below):
* [x] Merge of PR400  https://github.com/mpi-forum/mpi-standard/pull/400 - resolves MPI_MAX_FROM_GROUP_TAG vs. MPI_MAX_STRINGTAG_LEN issues
* [x] Merge of PR409  https://github.com/mpi-forum/mpi-standard/pull/409 - resolves most of the MPI-T callback prototype issues
* [x] Merge of PR423  https://github.com/mpi-forum/mpi-standard/pull/423 - removal of PMPI_T_... references in the function index
* [x] Merge of PR429  https://github.com/mpi-forum/mpi-standard/pull/429 - backward incompatibility and change-log text for modification of MPI_T_CATEGORY_CHANGED
* [x] Adding change-log and backward incompatibility entry for change in PR429 
* [x] Page 782 of update 7 and page 788 line 22 in [mpi40-report_Issue405_PR449_v03.pdf](https://github.com/mpi-forum/mpi-issues/files/5770001/mpi40-report_Issue405_PR449_v03.pdf): Is the formatting with a 8pt longer line acceptable?
* [x] A.3, A.4, A.5: Missing Hyperrefs to the primary locations of the functions and types (see also comments in [PR413]( https://github.com/mpi-forum/mpi-standard/pull/413) and Issue https://github.com/mpi-forum/mpi-issues/issues/422)
  * [x] for the function names, see Issue #422 and [PR476](https://github.com/mpi-forum/mpi-standard/pull/476)
  * [x] for all type declarations
* [x] Verfication whether the changelog is complete, e.g., references all new functions, types, and constants correctly (done in [PR461](https://github.com/mpi-forum/mpi-standard/pull/461))
* [x] Checking whether all references in the MPI Indexes have a useful primary index:
  * [ ] Constants (--> may be postponed until MPI-4.1)
  * [x] Declarations (all Fortran types should additionally get the primary references of the C types: don by [PR486]( https://github.com/mpi-forum/mpi-standard/pull/486))
  * [x] Callbacks (done by [PR449]( https://github.com/mpi-forum/mpi-standard/pull/449)) 
  * [x] Functions witout _c
  * [x] Functions with _c (by [PR474]( https://github.com/mpi-forum/mpi-standard/pull/474))
* [x] Formatting problem "Linebreak before '*)'" in Fortran old-style binding of MPI_Waitall and MPI_Testall, reported in Issue [#364](https://github.com/mpi-forum/mpi-issues/issues/364#issuecomment-739013771)

See also "#-TODO" and "Done" comments in the latest version of the commented mpi-4-rc pdf file:
[mpi40-rc-nov-20_rab_Chap-8+16+17+A+B_update14.pdf](https://github.com/mpi-forum/mpi-issues/files/5869068/mpi40-rc-nov-20_rab_Chap-8%2B16%2B17%2BA%2BB_update14.pdf)

# References

- [mpi40-rc-nov-20_rab_Chap-8+16+17+A+B.pdf](https://github.com/mpi-forum/mpi-issues/files/5683474/mpi40-rc-nov-20_rab_Chap-8%2B16%2B17%2BA%2BB.pdf) as voted for rc-40 at the MPI forum meeting Dec 7-10, 2020
- ( [mpi40-rc-nov-20_rab_Chap-8+16+17+A+B_update1.pdf](https://github.com/mpi-forum/mpi-issues/files/5684583/mpi40-rc-nov-20_rab_Chap-8%2B16%2B17%2BA%2BB_update1.pdf) update 1 since rc-40 vote )
- ( [mpi40-rc-nov-20_rab_Chap-8+16+17+A+B_update2.pdf](https://github.com/mpi-forum/mpi-issues/files/5719015/mpi40-rc-nov-20_rab_Chap-8%2B16%2B17%2BA%2BB_update2.pdf) )
- ( [mpi40-rc-nov-20_rab_Chap-8+16+17+A+B_update3.pdf](https://github.com/mpi-forum/mpi-issues/files/5725734/mpi40-rc-nov-20_rab_Chap-8%2B16%2B17%2BA%2BB_update3.pdf) updated after PR 439 )
- ( [mpi40-rc-nov-20_rab_Chap-8+16+17+A+B_update4.pdf](https://github.com/mpi-forum/mpi-issues/files/5731874/mpi40-rc-nov-20_rab_Chap-8%2B16%2B17%2BA%2BB_update4.pdf) updated after PR 443 )
- ( [mpi40-rc-nov-20_rab_Chap-8+16+17+A+B_update5.pdf](https://github.com/mpi-forum/mpi-issues/files/5736752/mpi40-rc-nov-20_rab_Chap-8%2B16%2B17%2BA%2BB_update5.pdf) updated after PR 444 )
- ( [mpi40-rc-nov-20_rab_Chap-8+16+17+A+B_update7.pdf](https://github.com/mpi-forum/mpi-issues/files/5770078/mpi40-rc-nov-20_rab_Chap-8%2B16%2B17%2BA%2BB_update7.pdf) updated after PR 449 )
- ( [mpi40-rc-nov-20_rab_Chap-8+16+17+A+B_update8.pdf](https://github.com/mpi-forum/mpi-issues/files/5784440/mpi40-rc-nov-20_rab_Chap-8%2B16%2B17%2BA%2BB_update8.pdf) updated after PR 461 )
- ( [mpi40-rc-nov-20_rab_Chap-8+16+17+A+B_update9.pdf](https://github.com/mpi-forum/mpi-issues/files/5790411/mpi40-rc-nov-20_rab_Chap-8%2B16%2B17%2BA%2BB_update9.pdf) updated after last updates of PR449 and PR 461 )
- ( [mpi40-rc-nov-20_rab_Chap-8+16+17+A+B_update10.pdf](https://github.com/mpi-forum/mpi-issues/files/5795358/mpi40-rc-nov-20_rab_Chap-8%2B16%2B17%2BA%2BB_update10.pdf) )
- ( [mpi40-rc-nov-20_rab_Chap-8+16+17+A+B_update11.pdf](https://github.com/mpi-forum/mpi-issues/files/5797779/mpi40-rc-nov-20_rab_Chap-8%2B16%2B17%2BA%2BB_update11.pdf) updated after PR 486 )
- ( [mpi40-rc-nov-20_rab_Chap-8+16+17+A+B_update12.pdf](https://github.com/mpi-forum/mpi-issues/files/5814023/mpi40-rc-nov-20_rab_Chap-8%2B16%2B17%2BA%2BB_update12.pdf) after revisiting Issues #364, #366, and #375 + PR 497 )
- ( [mpi40-rc-nov-20_rab_Chap-8+16+17+A+B_update13.pdf](https://github.com/mpi-forum/mpi-issues/files/5842702/mpi40-rc-nov-20_rab_Chap-8%2B16%2B17%2BA%2BB_update13.pdf) after #432 and PR 493, 498, 499, 500, 501 )
- [mpi40-rc-nov-20_rab_Chap-8+16+17+A+B_update14.pdf](https://github.com/mpi-forum/mpi-issues/files/5869068/mpi40-rc-nov-20_rab_Chap-8%2B16%2B17%2BA%2BB_update14.pdf) after #332 and #349 with PR 504, 506, and 508.

- PR427 https://github.com/mpi-forum/mpi-standard/pull/427 Sorting of the changelog based on
[mpi40-rc-nov-20_Sort-changelog-B.1.2.xlsx](https://github.com/mpi-forum/mpi-issues/files/5684589/mpi40-rc-nov-20_Sort-changelog-B.1.2.xlsx)
and correction in chap-datatype that the Azure Pipelines CI works without failures.
- PR434 https://github.com/mpi-forum/mpi-standard/pull/434  Corrections to
  - Annex B. Change-log
  - (Chap 2. Terms: an additional subsection label)
  - Chap 19. Binding: Only Section 19.1.5, see change-log Items 1 and 33
  - Chap 18. Backward Compatibility. see change-log Item 33
- PR439 https://github.com/mpi-forum/mpi-standard/pull/439 Corrections to
  - Chap.8 Topologies 
  - Fixes for the MPI 4.0rc in the topology chapter brought up in
      - Issue mpi-forum/mpi-issues#343
      - Issue mpi-forum/mpi-issues#353
- PR443 https://github.com/mpi-forum/mpi-standard/pull/442 Corrections to
  - Chap.10 The Info Object
  - Fixes for the MPI 4.0rc in the "Info Object" chapter brought up in
      - Issue mpi-forum/mpi-issues#364 
      - and during the Dec. 2020 meeting
- PR442 https://github.com/mpi-forum/mpi-standard/pull/442 Corrections to
  - Chap.16 Deprecated Interfaces
  - Fixes for the MPI 4.0rc in the "Deprecated Interfaces" chapter brought up in
      - Issue mpi-forum/mpi-issues#366
      - and during the Dec. 2020 meeting
- PR441 https://github.com/mpi-forum/mpi-standard/pull/441 Corrections to
  - Chap.17 Removed Interfaces
  - Fixes for the MPI 4.0rc in the "Removed Interfaces" chapter brought up in
      - Issue mpi-forum/mpi-issues#367
      - and during the Dec. 2020 meeting
- PR444 https://github.com/mpi-forum/mpi-standard/pull/444 Corrections to
  - Annex A. Language Binding Summaries
  - Fixes for the MPI 4.0rc in the "Language Binding Summaries" annex A brought up in
      - Issue mpi-forum/mpi-issues#405 and others
      - and during the Dec. 2020 meeting
- PR449 https://github.com/mpi-forum/mpi-standard/pull/449 Corrections to
  - Callback Prototype Function Index
  - and some smaller ones in other chapters
  - Fixes for the MPI 4.0rc in the "Callback Prototype Function Index" annex A brought up in
      - Issue mpi-forum/mpi-issues#405 and others
      - and during the Dec. 2020 meeting
- PR461 https://github.com/mpi-forum/mpi-standard/pull/461 Corrections to
  - Change-log
  - and appropriate additions to all Index sections
  - Fixes for the MPI 4.0rc in the "Change-log" annex B brought up in
      - Issue mpi-forum/mpi-issues#405 and others
      - and during the Dec. 2020 meeting

**Comment by RolfRabenseifner** [comment]
The "no more 30 character limit for MPI identifiers" is done. See Sections 2.2, 18.1, 19.1.5 page 791 Table 19.1 2B and page 793f 2nd advice to impl., B.1.2 Items 1 and 33, in following pdf:
[mpi-standard-mpi-4-rc_2020-PR432+PR434+PR435.pdf](https://github.com/mpi-forum/mpi-issues/files/5719009/mpi-standard-mpi-4-rc_2020-PR432%2BPR434%2BPR435.pdf)
It also includes all rc-40 updates to the change-log chapter.

**Comment by RolfRabenseifner** [comment]
All changes by PR 439 are checked and verified in
[mpi40-report_Issue405_PR439.pdf](https://github.com/mpi-forum/mpi-issues/files/5725671/mpi40-report_Issue405_PR439.pdf)
PR 439 includes and superseeds the old PR 332, which will be therefore closed.

**Comment by RolfRabenseifner** [comment]
All changes by PR 441 are checked and verified in
[mpi40-report_Issue405_PR441.pdf](https://github.com/mpi-forum/mpi-issues/files/5730216/mpi40-report_Issue405_PR441.pdf)

**Comment by RolfRabenseifner** [comment]
All changes by PR 442 (deprecated.tex) are checked and verified in
[mpi40-report_Issue405_PR442.pdf](https://github.com/mpi-forum/mpi-issues/files/5731268/mpi40-report_Issue405_PR442.pdf)

**Comment by RolfRabenseifner** [comment]
All changes by PR 443 (misc-2.tex - the info chapter) are checked and verified in
[mpi40-report_Issue405_PR443.pdf](https://github.com/mpi-forum/mpi-issues/files/5731769/mpi40-report_Issue405_PR443.pdf)

**Comment by RolfRabenseifner** [comment]
All changes by PR 444 (appLang-Const.tex - the Annex A.1) are checked and verified in
[mpi40-report_Issue405_PR444.pdf](https://github.com/mpi-forum/mpi-issues/files/5736739/mpi40-report_Issue405_PR444.pdf)

**Comment by RolfRabenseifner** [comment]
All changes by PR 443 (misc-2.tex - the info chapter) are in
[mpi40-report_Issue405_PR449_v02.pdf](https://github.com/mpi-forum/mpi-issues/files/5768941/mpi40-report_Issue405_PR449_v02.pdf)

**Comment by RolfRabenseifner** [comment]
All changes by PR 449 (mainly callback references, including a late footnote number correction in terms) are checked and verified in
[mpi40-report_Issue405_PR449_v03.pdf](https://github.com/mpi-forum/mpi-issues/files/5770001/mpi40-report_Issue405_PR449_v03.pdf)

**Comment by RolfRabenseifner** [comment]
All changes by PR 461 (completing the change-log) are checked and verified in
[mpi40-report_Issue405_PR461_v05.pdf](https://github.com/mpi-forum/mpi-issues/files/5784425/mpi40-report_Issue405_PR461_v05.pdf)

It contains only small additional sentences, see pages 1040-1042. They are the basis for completing the index sections that nearly all new constants, types, callbacks and functions have a reference to the page of their change-log item.

**Comment by RolfRabenseifner** [comment]
Thank you Wesley for your help to fix the problem with github and binding-2.tex in PR449.
The corrected pdf is now:  
[mpi40-report_Issue405_PR449_v05.pdf](https://github.com/mpi-forum/mpi-issues/files/5788639/mpi40-report_Issue405_PR449_v05.pdf)

**Comment by RolfRabenseifner** [comment]
And after latest update of PR449: [mpi40-report_Issue405_PR449_v06.pdf](https://github.com/mpi-forum/mpi-issues/files/5790402/mpi40-report_Issue405_PR449_v06.pdf)

**Comment by RolfRabenseifner** [comment]
Latest commented pdf of PR461: 
[mpi40-report_Issue405_PR461_v06.pdf](https://github.com/mpi-forum/mpi-issues/files/5790427/mpi40-report_Issue405_PR461_v06.pdf) (does not include the changes in PR449)

**Comment by RolfRabenseifner** [comment]
Latest pdf of PR449: 
[mpi40-report_Issue405_PR449_v07.pdf](https://github.com/mpi-forum/mpi-issues/files/5794797/mpi40-report_Issue405_PR449_v07.pdf) (does not include changes in PR461)

**Comment by RolfRabenseifner** [comment]
Latest pdf of PR461 (change-log updates, now also including merged PR449, i.e., complete callback index and others):
[mpi40-report_Issue405_PR461_v10_NoComments.pdf](https://github.com/mpi-forum/mpi-issues/files/5797574/mpi40-report_Issue405_PR461_v10_NoComments.pdf)

**Comment by RolfRabenseifner** [comment]
All \`\`\_c'' in \code{...} - latest pdf of PR499: 
[mpi40-report_Issue405_PR499_v09.pdf](https://github.com/mpi-forum/mpi-issues/files/5825747/mpi40-report_Issue405_PR499_v09.pdf)

**Comment by RolfRabenseifner** [comment]
Latest pdf for PR493: [mpi40-report_Issue405_PR493_v04.pdf](https://github.com/mpi-forum/mpi-issues/files/5837851/mpi40-report_Issue405_PR493_v04.pdf)

**Comment by wgropp** [comment]
Is there a PR for any remaining issues? If not, can this be closed?

**Comment by RolfRabenseifner** [comment]
> Is there a PR for any remaining issues? If not, can this be closed?

@wgropp I'm not sure, because https://github.com/mpi-forum/mpi-standard/pull/476 did not show a pdf. Therefore, I could not check whether the MPI_Xxxx declarations in A.3 and A.4 are now again with hyperrefs to their main (i.e. first underlined) page in the Declarations Index. 
Therefore the checkbox  
- A.3, A.4, A.5: Missing Hyperrefs to the primary locations of the functions and types (see also comments in PR413 and Issue #422)
 for the function names, see Issue #422 and PR476
  * [ ] for all type declarations

is still not checked. 
This is the last remaining checkbox before I can close this issue. (Independent of that, I still need to update my report of changes.)

**Comment by wgropp** [comment]
I believe these links are present; I see them for both the functions and MPI types.

**Comment by RolfRabenseifner** [comment]
All pull requests are merged. Therefore, I close this issue 405.

**Comment by RolfRabenseifner** [comment]
Reopened for the PRs from the Feb 2021 meeting, see first part of the description of this issue.

**Comment by wesbland** [comment]
@RolfRabenseifner I believe all of the tasks remaining in this issue have moved into their own issues so I'm going to close this. If that's not true, can you create issues for anything left?
